app/main.py

```diff
@@ -86,128 +92,48 @@
-    for a in red_knight["armour"]:
-        red_knight["protection"] += a["protection"]
-
-    # apply weapon
-    red_knight["power"] += red_knight["weapon"]["power"]
-
-    # apply potion if exist
-    if red_knight["potion"] is not None:
-        if "power" in red_knight["potion"]["effect"]:
-            red_knight["power"] += red_knight["potion"]["effect"]["power"]
-
-        if "protection" in red_knight["potion"]["effect"]:
-            red_knight["protection"] += red_knight["potion"]["effect"]["protection"]
-
-        if "hp" in red_knight["potion"]["effect"]:
-            red_knight["hp"] += red_knight["potion"]["effect"]["hp"]
-
-    # -------------------------------------------------------------------------------
-    # BATTLE:
-
-    # 1 Lancelot vs Mordred:
-    lancelot["hp"] -= mordred["power"] - lancelot["protection"]
-    mordred["hp"] -= lancelot["power"] - mordred["protection"]
-
-    # check if someone fell in battle
-    if lancelot["hp"] <= 0:
-        lancelot["hp"] = 0
-
-    if mordred["hp"] <= 0:
-        mordred["hp"] = 0
-
-    # 2 Arthur vs Red Knight:
-    arthur["hp"] -= red_knight["power"] - arthur["protection"]
-    red_knight["hp"] -= arthur["power"] - red_knight["protection"]
-
-    # check if someone fell in battle
-    if arthur["hp"] <= 0:
-        arthur["hp"] = 0
-
-    if red_knight["hp"] <= 0:
-        red_knight["hp"] = 0
-
-    # Return battle results:
+def knights_fight(first: Knight, second: Knight) -> None:
+    first.get_damage(second)
+    second.get_damage(first)
+
+
+def battle(knights_config: dict) -> dict:
+    all_knights = []
+    for knight in knights_config:
+        temp_knight = Knight(
+            name=knights_config[knight]["name"],
+            power=knights_config[knight]["power"],
+            hp=knights_config[knight]["hp"]
+        )
+        for i in range(len(knights_config[knight]["armour"])):
+            temp_knight.set_armour(Armour(
+                knights_config[knight]["armour"][i]["part"],
+                protection=knights_config[knight]["armour"][i]["protection"]
+            ))
+        temp_knight.set_weapon(Weapon(
+            knights_config[knight]["weapon"]["name"],
+            power=knights_config[knight]["weapon"]["power"]
+        ))
+        if knights_config[knight]["potion"]:
+            temp_potion = Potion(knights_config[knight]["potion"]["name"])
+            if "power" in knights_config[knight]["potion"]["effect"]:
+                temp_potion.power += (
+                    knights_config[knight]["potion"]["effect"]["power"]
+                )
+            if "protection" in knights_config[knight]["potion"]["effect"]:
+                temp_potion.protection += (
+                    knights_config[knight]["potion"]["effect"]["protection"]
+                )
+            if "hp" in knights_config[knight]["potion"]["effect"]:
+                temp_potion.hp += (
+                    knights_config[knight]["potion"]["effect"]["hp"]
+                )
```

Comment:
You can define variable like "potion_data" to not repeat the knights_config[knight]["potion"] code